mlflow/azureml/__init__.py

```diff
@@ -206,8 +207,163 @@ def build_image(model_uri, workspace, image_name=None, model_name=None,
+    model_python_version = model_pyfunc_conf.get(pyfunc.PY_VERSION, None)
+    if model_python_version is not None and\
+            StrictVersion(model_python_version) < StrictVersion("3.0.0"):
+        raise MlflowException(
+                message=("Azure ML can only deploy models trained in Python 3 and above. See"
+                         " the following MLflow GitHub issue for a thorough explanation of this"
+                         " limitation and a workaround to enable support for deploying models"
+                         " trained in Python 2: https://github.com/mlflow/mlflow/issues/668"),
+                error_code=INVALID_PARAMETER_VALUE)
+
+    tags = _build_tags(model_uri=model_uri, model_python_version=model_python_version)
+
+    if service_name is None:
+        service_name = _get_mlflow_azure_name_from_run(run_id)
+    if model_name is None:
+        model_name = _get_mlflow_azure_name_from_run(run_id)
+
+    with TempDir(chdr=True) as tmp:
+        model_directory_path = tmp.path("model")
+        tmp_model_path = os.path.join(
+            model_directory_path,
+            _copy_file_or_tree(src=absolute_model_path, dst=model_directory_path))
+
+
+        registered_model = AzureModel.register(workspace=workspace, model_path=tmp_model_path,
+                                               model_name=model_name, tags=tags)
+
+        _logger.info("Registered an Azure Model with name: `%s` and version: `%s`",
+                     registered_model.name, registered_model.version)
+
+        # Create an execution script (entry point) for the image's model server. Azure ML requires
+        # the container's execution script to be located in the current working directory during
+        # image creation, so we create the execution script as a temporary file in the current
+        # working directory.
+        execution_script_path = tmp.path("execution_script.py")
+        _create_execution_script(output_path=execution_script_path, azure_model=registered_model)
+
+        environment = None
+        if pyfunc.ENV in model_pyfunc_conf:
+            environment = AzureEnvironment.from_conda_specification(_get_mlflow_azure_name_from_run(run_id), os.path.join(tmp_model_path, model_pyfunc_conf[pyfunc.ENV]))
+        else:
+            raise MlflowException("Cannot deploy pyfunc model without an environment to Azure ML.")
+
+        if mlflow_home is not None:
+            wheel = _create_mlflow_wheel(mlflow_home, tmp.path("dist"))
+            whl_url = AzureEnvironment.add_private_pip_wheel(workspace=workspace, file_path = wheel, exist_ok=True)
+            environment.python.conda_dependencies.add_pip_package(whl_url)
+        else:
+            environment.python.conda_dependencies.add_pip_package("mlflow~={}".format(mlflow_version))
+
+        environment.python.conda_dependencies.add_pip_package("azureml-defaults~={}".format(AZUREML_VERSION))
```

Comment:
Can we add a comment explaining why installing `azureml-defaults` is necessary? The description from https://pypi.org/project/azureml-defaults/ is pretty vague:

```
The azureml-defaults is a metapackage that is used internally by Azure Machine Learning.
```

Reply:
It installs the flask serve stack into the image - poorly named, we're working on it